### ISSUE TITLE
Some confusion about the resources

### ISSUE THREAD

**Issue by TheDonsky** [kcat/openal-soft#580]
Hello,

I have a few questions about resources:

0. Are audio buffers shared by all contexts, tied to the device? (Probably the answer is yes, since it worked on my PC, but since the API demands a context to be bound during creation, I was not sure, if I was simply lucky with this implementation or it was a part of the standard);
1. I can create more than one context and cycle through all, creating duplicate sources to emulate "multiple listeners" if my game demands that, right? (Again, works on my device, so probably yes, but alternatively, I can simply create a single context and use relative positions on sources, duplicating them to produce the same effect)
2. I used alcGetIntegerv to retrieve ALC_MONO_SOURCES and ALC_STEREO_SOURCES from my device, however, I have a bit of a hard time understanding if these source limits correspond to the number of sources I can safety create, or play (likely play, but it's not clear to me just yet). Also, I want to know if these numbers are values per context or per device. (matters for multiple listeners)
3. In case ALC_MONO_SOURCES  corresponds to the actively playing source count and I exceed it, will the API automatically ignore some of the sources and leave others, or should I manually manage those numbers and create some kind of an eviction mechanism?
4. ALC_STEREO_SOURCES returned 1; Does this mean I can not play more than a single background sound at a time? (I can "hack" the system, by splitting the audio channels into separate mono sources and placing them in listener-space positions, but I would rather not do that)
5. We have only MONO and STEREO formats, so does that mean we can not natively play something like a 5.1 encoded audio? (Again, I can split the channels and place them in the world, creating a "solution", but that's a bit of a hack and I'm willing to avoid it if I can).

PS: I know, maybe, GitHub issues are not necessarily a place to ask such questions, but this is the closest thing to support I could think of and some repositories actually use it as such and maybe you do too, so sorry if this is not the place

**Comment by kcat** [comment]
> Are audio buffers shared by all contexts, tied to the device?

Yes, buffers are stored on the device and can be accessed through any context on that device. A context needs to be set to generate and use them, but the buffers are recognized according to the device the context belongs to, rather than the specific context.

> I can create more than one context and cycle through all, creating duplicate sources to emulate "multiple listeners" if my game demands that, right?

With OpenAL Soft, yes. I don't know how many other implementations support creating more than one context at a time, but it's certainly possible for an implementation to.

> I used alcGetIntegerv to retrieve ALC_MONO_SOURCES and ALC_STEREO_SOURCES from my device, however, I have a bit of a hard time understanding if these source limits correspond to the number of sources I can safety create, or play (likely play, but it's not clear to me just yet). Also, I want to know if these numbers are values per context or per device. (matters for multiple listeners)

Going by the spec, an implementation can fail to set a buffer on a source beyond the limit (e.g. `alSourcei(source, AL_BUFFER, buffer)` and `alSourceQueueBuffers(...)` would fail), or it can allow them to set a buffer but fail to play. Or it can continue to allow them to play, until some other internal limit is reached. You're guaranteed to be able to create, set, and play the number of sources specified by those limits, but it's implementation-defined behavior what happens when you exceed them (e.g. if `ALC_MONO_SOURCES` if 254 and `ALC_STEREO_SOURCES` is 2, and you manage to play 255 mono sources, you may or may not be able to play 1 or 2 stereo sources; or if you play 3 stereo sources, you may or may not still be able to play all 254 mono sources).

As far as OpenAL Soft specifically is concerned, you can create and play the number of sources determined by the sum of `ALC_MONO_SOURCES` and `ALC_STEREO_SOURCES` however you want. It won't allow you to create more than that, but once you have a source it makes no difference if it plays mono, stereo, quad, or anything else. If you have a source, you can play it.

The spec is very not clear on whether the limit applies per-context or per-device, it only mentions those attributes can be specified as a hint during context creation (along with `ALC_FREQUENCY` and some others), and then says they can be later queried (from the device) to see what's really set. The actual values are set per-device (I don't see how it can be otherwise, since it can only be queried using a device instead of a context), but it doesn't say whether the limit imposed is on the sources in the created context, or on the sources in all contexts on the device.

For OpenAL Soft, the limit is handled per-context. It merely stops allowing sources to be created after reaching the sum count on a given context, but once you have a source, you can use it.

> In case ALC_MONO_SOURCES corresponds to the actively playing source count and I exceed it, will the API automatically ignore some of the sources and leave others, or should I manually manage those numbers and create some kind of an eviction mechanism?

I'm not aware of any implementation that does any virtualization like that. If some limit is exceeded, it generates a failure, I wouldn't expect them to silently ignore calls to play or silence some active sources to play others. I could be wrong, but it's certainly not the case with OpenAL Soft.

> ALC_STEREO_SOURCES returned 1; Does this mean I can not play more than a single background sound at a time?

Technically sort of. If you try to play more than one stereo source, there's no real saying what'll happen, so it wouldn't be safe to play more than one in that case. However, with OpenAL Soft it's not that restrictive; if `ALC_MONO_SOURCES` is 255 and `ALC_STEREO_SOURCES` is 1, you can play 256 stereo sources no problem (obviously you'll incur the CPU cost of mixing 2 channels per source for stereo sounds compared to 1 channel per source for mono sounds, but it'll do as you tell it to).

> We have only MONO and STEREO formats, so does that mean we can not natively play something like a 5.1 encoded audio?

The `AL_EXT_MCFORMATS` extension provides 5.1 buffer formats (`AL_FORMAT_51CHN8` for unsigned 8-bit, `AL_FORMAT_51CHN16` for signed 16-bit, and `AL_FORMAT_51CHN32` for 32-bit float samples). It also provides Quad, 6.1, and 7.1 formats. The `AL_EXT_BFORMAT` extension provides formats for 2D and 3D B-Format (aka ambisonics or soundfield) formats, too. All supported by OpenAL Soft.

> PS: I know, maybe, GitHub issues are not necessarily a place to ask such questions, but this is the closest thing to support I could think of and some repositories actually use it as such and maybe you do too, so sorry if this is not the place

It's perfectly fine to ask here, don't worry. It's something other people can find if they have similar questions, and if a question spawns a feature request or idea, it's a good place to keep track of it.

**Comment by TheDonsky** [comment]
Thank you
